File: packages/dvai-bridge-dotnet/src/DVAIBridge/Pairing/PairingHandshake.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace DVAIBridge.Pairing;

/// <summary>
/// HMAC-SHA256 helpers for the LAN-pairing handshake. Mirrors the TS
/// implementation in <c>@dvai-bridge/core/src/pairing/handshake.ts</c>:
/// canonical message format, base64-url encoding, lowercase-hex SHA-256
/// body hashing.
///
/// <para>
/// The first time Device A wants to offload to Device B over the LAN, A
/// POSTs <c>/v1/dvai/handshake</c> to B with its identity + a fresh
/// nonce. B surfaces a UI prompt to the user; on approve, B generates a
/// 256-bit pairing key and returns it. From then on, A includes
/// <c>X-DVAI-Pairing: HMAC-SHA256(pairingKey, canonicalMessage)</c> on
/// every offload request to B.
/// </para>
/// </summary>
public static class PairingHandshake
{
    private const string EmptyBodyHash =
        "0000000000000000000000000000000000000000000000000000000000000000";

    /// <summary>Generate a fresh 256-bit pairing key (base64-url encoded).</summary>
    public static string GeneratePairingKey()
    {
        var bytes = new byte[32];
        RandomNumberGenerator.Fill(bytes);
        return Base64UrlEncode(bytes);
    }

    /// <summary>Generate a fresh nonce (128 bits, base64-url encoded).</summary>
    public static string GenerateNonce()
    {
        var bytes = new byte[16];
        RandomNumberGenerator.Fill(bytes);
        return Base64UrlEncode(bytes);
    }

    /// <summary>
    /// Compose the canonical message that gets HMAC-signed for a peer-to-peer
    /// offload request. The peer recomputes the same string and verifies.
    ///
    /// Format: <c>{nonce}\n{METHOD}\n{path}\n{bodyHash}</c> — bodyHash is
    /// the lowercase-hex SHA-256 of the request body bytes (or 64 zeros
    /// for an empty body).
    /// </summary>
    public static string ComposeSignedMessage(string nonce, string method, string path, string? body)
    {
        if (nonce is null) throw new ArgumentNullException(nameof(nonce));
        if (method is null) throw new ArgumentNullException(nameof(method));
        if (path is null) throw new ArgumentNullException(nameof(path));

        var bodyHash = string.IsNullOrEmpty(body) ? EmptyBodyHash : Sha256Hex(body!);
        return $"{nonce}\n{method.ToUpperInvariant()}\n{path}\n{bodyHash}";
    }

    /// <summary>
    /// HMAC-SHA256(pairingKey, message). Returns the base64-url encoded
    /// signature suitable for the <c>X-DVAI-Pairing</c> header.
    /// </summary>
    public static string SignHmac(string pairingKey, string message)
    {
        if (pairingKey is null) throw new ArgumentNullException(nameof(pairingKey));
        if (message is null) throw new ArgumentNullException(nameof(message));

        var keyBytes = Base64UrlDecode(pairingKey);
        using var hmac = new HMACSHA256(keyBytes);
        var sig = hmac.ComputeHash(Encoding.UTF8.GetBytes(message));
        return Base64UrlEncode(sig);
    }

    /// <summary>
    /// Verify an HMAC. Constant-time comparison after recomputing the
    /// expected signature.
    /// </summary>
    public static bool VerifyHmac(string pairingKey, string message, string signature)
    {
        var expected = SignHmac(pairingKey, message);
        return CryptographicOperations.FixedTimeEquals(
            Encoding.ASCII.GetBytes(expected),
            Encoding.ASCII.GetBytes(signature));
    }

    private static string Sha256Hex(string input)
    {
        var bytes = Encoding.UTF8.GetBytes(input);
        var hash = SHA256.HashData(bytes);
        var sb = new StringBuilder(hash.Length * 2);
        foreach (var b in hash) sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    private static string Base64UrlEncode(byte[] bytes)
    {
        var b64 = Convert.ToBase64String(bytes);
        var sb = new StringBuilder(b64.Length);
        foreach (var ch in b64)
        {
            switch (ch)
            {
                case '+': sb.Append('-'); break;
                case '/': sb.Append('_'); break;
                case '=': break;
                default: sb.Append(ch); break;
            }
        }
        return sb.ToString();
    }

    private static byte[] Base64UrlDecode(string s)
    {
        var b64 = s.Replace('-', '+').Replace('_', '/');
        var pad = (4 - (b64.Length % 4)) % 4;
        if (pad > 0) b64 += new string('=', pad);
        return Convert.FromBase64String(b64);
    }
}

File: packages/dvai-bridge-android-shared-core/android/src/main/java/co/deepvoiceai/bridge/shared/core/pairing/PairingHandshake.kt
package co.deepvoiceai.bridge.shared.core.pairing

import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * LAN-pairing handshake helpers (Kotlin mirror of
 * `@dvai-bridge/core/src/pairing/handshake.ts`).
 *
 * The first time Device A wants to offload to Device B over the LAN,
 * A POSTs `/v1/dvai/handshake` to B with its identity + a nonce. B
 * surfaces a UI prompt to the user; on approve, B generates a 256-bit
 * pairing key and returns it. From then on, A includes
 * `X-DVAI-Pairing: HMAC-SHA256(pairingKey, body)` on every offload
 * request to B.
 *
 * All keys + signatures use base64-url encoding, no padding.
 *
 * Implementation note: HMAC is via `javax.crypto.Mac` — we don't pull
 * in BouncyCastle since AndroidKeyStore + the platform Mac provider
 * already cover SHA-256 on every supported API level (24+).
 */
object PairingHandshake {
    /** Generate a fresh 256-bit pairing key (base64-url encoded). */
    fun generatePairingKey(): String {
        val bytes = ByteArray(32)
        SecureRandom().nextBytes(bytes)
        return base64UrlEncode(bytes)
    }

    /** Generate a fresh 128-bit nonce for a handshake request. */
    fun generateNonce(): String {
        val bytes = ByteArray(16)
        SecureRandom().nextBytes(bytes)
        return base64UrlEncode(bytes)
    }

    /**
     * HMAC-SHA256(key, message). Used to sign offload requests so the
     * peer can verify they came from a paired device.
     */
    fun signHmac(pairingKey: String, message: String): String {
        val keyBytes = decodeBase64UrlBytes(pairingKey)
        val mac = Mac.getInstance("HmacSHA256")
        mac.init(SecretKeySpec(keyBytes, "HmacSHA256"))
        val sig = mac.doFinal(message.toByteArray(Charsets.UTF_8))
        return base64UrlEncode(sig)
    }

    /** Verify an HMAC. Returns true on match (constant-time-ish). */
    fun verifyHmac(pairingKey: String, message: String, signature: String): Boolean {
        val expected = signHmac(pairingKey, message)
        return constantTimeEquals(expected, signature)
    }

    /**
     * Compose the canonical message that gets HMAC-signed for a
     * peer-to-peer offload request. The peer recomputes the same
     * string and verifies.
     *
     * Format: `${nonce}\n${method}\n${path}\n${bodyHash}` — bodyHash
     * is the hex-encoded SHA-256 of the request body bytes (or 64
     * zero-bytes for a missing body). Method is uppercased for
     * canonicalization.
     */
    fun composeSignedMessage(
        nonce: String,
        method: String,
        path: String,
        body: String?,
    ): String {
        val bodyHash = if (body != null) sha256Hex(body) else "0".repeat(64)
        return "$nonce\n${method.uppercase()}\n$path\n$bodyHash"
    }

    private fun sha256Hex(input: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val bytes = digest.digest(input.toByteArray(Charsets.UTF_8))
        val sb = StringBuilder(bytes.size * 2)
        for (b in bytes) {
            val v = b.toInt() and 0xff
            sb.append(HEX[v shr 4]).append(HEX[v and 0x0f])
        }
        return sb.toString()
    }

    private val HEX = "0123456789abcdef".toCharArray()

    internal fun base64UrlEncode(bytes: ByteArray): String {
        val sb = StringBuilder()
        val b64 = java.util.Base64.getEncoder().encodeToString(bytes)
        for (c in b64) {
            when (c) {
                '+' -> sb.append('-')
                '/' -> sb.append('_')
                '=' -> Unit
                else -> sb.append(c)
            }
        }
        return sb.toString()
    }

    internal fun decodeBase64UrlBytes(s: String): ByteArray {
        val sb = StringBuilder(s.length + 4)
        for (c in s) {
            when (c) {
                '-' -> sb.append('+')
                '_' -> sb.append('/')
                else -> sb.append(c)
            }
        }
        while (sb.length % 4 != 0) sb.append('=')
        return java.util.Base64.getDecoder().decode(sb.toString())
    }

    private fun constantTimeEquals(a: String, b: String): Boolean {
        if (a.length != b.length) return false
        var diff = 0
        for (i in a.indices) {
            diff = diff or (a[i].code xor b[i].code)
        }
        return diff == 0
    }
}
